igvc_platform/src/imu/YostLabDriver.cpp

```diff
@@ -1,39 +1,81 @@
 #include "YostLabDriver.h"
 #include <parameter_assertions/assertions.h>
 
-YostLabDriver::YostLabDriver(ros::NodeHandle& nh_, ros::NodeHandle& priv_nh_)
+YostLabDriver::YostLabDriver(ros::NodeHandle &nh_, ros::NodeHandle &priv_nh_)
   : SerialInterface(priv_nh_), yostlab_priv_nh_(priv_nh_), yostlab_nh_(nh_)
 {
-  this->SerialConnect();
-  this->imu_pub_ = this->yostlab_nh_.advertise<sensor_msgs::Imu>("/imu", 10);
-  // use identity matrix as default orientation correction
+  serialConnect();
+  imu_pub_ = yostlab_nh_.advertise<sensor_msgs::Imu>("/imu", 10);
+  updater.setHardwareIDf("IMU: %s", getSerialPort().c_str());
+  updater.add("IMU Diagnostic", this, &YostLabDriver::imu_diagnostic);
 
-  assertions::param(this->yostlab_priv_nh_, "imu_orientation_correction", this->imu_orientation_correction_,
+  // use identity matrix as default orientation correction
+  assertions::param(yostlab_priv_nh_, "imu_orientation_correction", imu_orientation_correction_,
                     std::vector<double>{ 1, 0, 0, 0, 1, 0, 0, 0, 1 });
   assertions::param(yostlab_priv_nh_, "orientation_rotation", orientation_rotation_, 0.0);
-  assertions::getParam(this->yostlab_priv_nh_, "frame_id", this->frame_id_);
+  assertions::getParam(yostlab_priv_nh_, "frame_id", frame_id_);
 }
 
 //! Destructor
 YostLabDriver::~YostLabDriver()
 {
+  updater.broadcast(diagnostic_msgs::DiagnosticStatus::ERROR, "IMU Node Terminated");
 }
+
 void YostLabDriver::restoreFactorySettings()
 {
-  this->SerialWriteString(RESTORE_FACTORY_SETTINGS);
+  serialWriteString(RESTORE_FACTORY_SETTINGS);
 }
-const std::string YostLabDriver::getSoftwareVersion()
+
+std::string YostLabDriver::getSoftwareVersion()
 {
-  this->SerialWriteString(GET_FIRMWARE_VERSION_STRING);
-  const std::string buf = this->SerialReadLine();
-  ROS_INFO_STREAM(this->logger << "Software version: " << buf);
+  flush();
+  serialWriteString(GET_FIRMWARE_VERSION_STRING);
+  const std::string buf = serialReadLine();
+  ROS_INFO_STREAM(logger << "Software version: " << buf);
   return buf;
 }
 
-const std::string YostLabDriver::getEulerDecomp()
+void YostLabDriver::imu_diagnostic(diagnostic_updater::DiagnosticStatusWrapper &stat)
 {
-  this->SerialWriteString(GET_EULER_DECOMPOSTION_ORDER);
-  const std::string buf = this->SerialReadLine();
+  if (sensor_temp_ > MAX_IMU_TEMP)
+  {
+    stat.summary(diagnostic_msgs::DiagnosticStatus::WARN, "IMU temp too high");
+  }
+  else if (sensor_temp_ < MIN_IMU_TEMP)
+  {
+    stat.summary(diagnostic_msgs::DiagnosticStatus::WARN, "IMU temp too low");
+  }
+  else if ((ros::Time::now() - lastUpdateTime_).toSec() > IMU_TIMEOUT_DELAY)
+  {
+    stat.summary(diagnostic_msgs::DiagnosticStatus::ERROR, "IMU not updating");
+  }
+  else if (abs(quaternion_length_ - 1.0) > QUATERNION_LENGTH_TOL)
+  {
+    stat.summary(diagnostic_msgs::DiagnosticStatus::WARN, "IMU quaternion isn't normalized");
+  }
+  else
+  {
+    stat.summary(diagnostic_msgs::DiagnosticStatus::OK, "IMU Online");
+  }
```

Comment:
I just noticed this - the order of this should be in terms of most severe to least severe, ie. I want to know that the IMU is not updating and not that the imu temp is too low or something.